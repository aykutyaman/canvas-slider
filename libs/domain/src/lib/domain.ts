export type Canvas = CanvasRenderingContext2D;

export type Image = {
  element: HTMLImageElement;
  sx: number;
  sy: number;
  sWidth: number;
  sHeight: number;
  dx: number;
  dy: number;
  dWidth: number;
  dHeight: number;
};

export type Images = {
  images: Array<Image>;
};

export type Loading = {
  kind: 'Loading';
};

export type Idle = {
  kind: 'Idle';
} & Images;

export type Dragging = {
  kind: 'Dragging';
} & Images;

export type State = Loading | Idle | Dragging;

export const initialState: State = {
  kind: 'Loading',
};

// actions
type MouseDown = {
  kind: 'MouseDown';
  payload: {
    event: MouseEvent;
    canvas: Canvas;
  };
};

type MouseUp = {
  kind: 'MouseUp';
  payload: {
    event: MouseEvent;
    canvas: Canvas;
  };
};

type MouseMove = {
  kind: 'MouseMove';
  payload: {
    event: MouseEvent;
    canvas: Canvas;
  };
};

export type ImagesPayload = {
  images: Array<HTMLImageElement>;
  canvas: Canvas;
};

export type ImagesLoaded = {
  kind: 'ImagesLoaded';
  payload: ImagesPayload;
};

export type Action = MouseDown | MouseUp | MouseMove | ImagesLoaded;

// Action creators
export type MousePayload = {
  event: MouseEvent;
  canvas: Canvas;
};
export const mouseDown = (payload: MousePayload): MouseDown => ({
  kind: 'MouseDown',
  payload,
});
export const mouseUp = (payload: MousePayload): MouseUp => ({
  kind: 'MouseUp',
  payload,
});
export const mouseMove = (payload: MousePayload): MouseMove => ({
  kind: 'MouseMove',
  payload,
});
export const imagesLoaded = (payload: ImagesPayload): ImagesLoaded => ({
  kind: 'ImagesLoaded',
  payload,
});
