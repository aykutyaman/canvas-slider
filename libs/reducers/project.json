{
  "name": "reducers",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/reducers/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/reducers/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/reducers/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
