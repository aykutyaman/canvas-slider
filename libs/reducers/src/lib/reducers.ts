import * as D from '@publitas/domain';

const assertNever = (value: never): never => {
  throw new Error('This value should be nnreacable' + value);
};
const noop = (state: D.State, _action: D.Action): D.State => {
  // console.log(`noop for: ${state.type} - ${_action.type}`);
  return state;
};

export const loadingImagesLoaded = (
  _state: D.Loading,
  action: D.ImagesLoaded
): D.Idle => {
  const ctx = action.payload.canvas; // ctx.canvas;
  const canvas = ctx.canvas;
  const images: Array<D.Image> = action.payload.images.map((img, i) => {
    const hRatio = canvas.width / img.width;
    const vRatio = canvas.height / img.height;
    const ratio = Math.min(hRatio, vRatio);
    const centerShift_x = (canvas.width - img.width * ratio) / 2;
    const centerShift_y = (canvas.height - img.height * ratio) / 2;

    return {
      element: img,
      sx: 0,
      sy: 0,
      sWidth: img.width,
      sHeight: img.height,
      dx: centerShift_x + i * canvas.width,
      dy: centerShift_y,
      dWidth: img.width * ratio,
      dHeight: img.height * ratio,
    };
  });

  return {
    kind: 'Idle',
    images,
  };
};

export const reducer = (state: D.State, action: D.Action): D.State => {
  switch (state.kind) {
    case 'Loading':
      switch (action.kind) {
        case 'ImagesLoaded':
          return loadingImagesLoaded(state, action);
        case 'MouseDown':
        case 'MouseUp':
        case 'MouseMove':
          return noop(state, action);
        default:
          return assertNever(action);
      }
    case 'Idle':
      switch (action.kind) {
        case 'ImagesLoaded':
        case 'MouseDown':
        case 'MouseUp':
        case 'MouseMove':
          return noop(state, action);
        default:
          return assertNever(action);
      }
    case 'Dragging':
      switch (action.kind) {
        case 'MouseDown':
        case 'MouseUp':
        case 'MouseMove':
        case 'ImagesLoaded':
          return noop(state, action);
        default:
          return assertNever(action);
      }
    default:
      return assertNever(state);
  }
};
