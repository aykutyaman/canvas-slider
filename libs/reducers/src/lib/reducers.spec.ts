import * as D from '@publitas/domain';
import { reducer } from './reducers';

const canvas = {
  canvas: {
    width: 640,
    height: 400,
  },
} as D.Canvas;

const images = [
  {
    width: 2880,
    height: 1800,
  },
  {
    width: 1500,
    height: 750,
  },
  { width: 1024, height: 768 },
  { width: 853, height: 1280 },
] as Array<HTMLImageElement>;

describe('reducers', () => {
  it('Loading-ImagesLoaded-Idle', () => {
    const given: D.Idle = {
      images: [],
      kind: 'Idle',
    };
    const when: D.ImagesLoaded = {
      kind: 'ImagesLoaded',
      payload: {
        canvas,
        images,
      },
    };
    expect(reducer(given, when)).toMatchSnapshot();
  });
});
