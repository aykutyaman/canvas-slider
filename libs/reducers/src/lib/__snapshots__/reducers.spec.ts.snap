// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers Loading-ImagesLoaded-Idle 1`] = `
Object {
  "images": Array [],
  "kind": "Idle",
}
`;
